### DISCUSSION TITLE
deepmd pair style not present after building

### DISCUSSION THREAD

**Discussion by magnamad** [deepmodeling/deepmd-kit#2954] [Q&A]
When I run the command ,` /home/madhavan/local/bin/lmp -h ` , I get the following message followed by usual help lines.

```
/home/madhavan/local/bin/lmp: Relink `/home/madhavan/anaconda3/envs/lpd_env/lib/python3.11/site-packages/tensorflow/libtensorflow_framework.so.2' with `/lib/x86_64-linux-gnu/libz.so.1' for IFUNC symbol `crc32_z'

``` 
Additionally , no deepmd pairstyle present when I list the styles using , /home/madhavan/local/bin/lmp -h.  I doubled checked by runnning a small lammps simulation  (after loading the plugin 
`pluigin load /home/madhavan/Softwares/deepmd-kit/lb/libdeepmd_lmp.so` )  and an error saying unknown command plugin load.
I also tried setting the the LAMMPS_PLUGIN_PATH environment variable and run the small lammps simulation . This time, it showed no deepmd pair style.

------

I build from source , deepmd kit (2.2.6) and lammps ( 2023 Aug ). Here are the installation steps I followed , please guide me to the source of the mistake.

`conda create --name lpd_env python=3.11`
conda activate lpd_env
`
pip install --upgrade tensorflow-cpu  # cpu only version
`
The installed tensorflow works

`
git clone https://github.com/deepmodeling/deepmd-kit.git deepmd-kit
`

`conda install mpi4py -c defaults -c conda-forge
HOROVOD_WITHOUT_GLOO=1 HOROVOD_WITH_TENSORFLOW=1 pip install horovod 
`
horovodrun --check-build works

`
cd deepmd-kit
pip install .
`

and dp -h  works.

Now for building the C++ interface

```

cmake -DUSE_TF_PYTHON_LIBS=TRUE -DCMAKE_INSTALL_PREFIX=$deepmd_root -DLAMMPS_SOURCE_ROOT=$lammps_source_code_path

make -j8
make install
```

Now in lammps build directory ,

```
cmake -D PKG_PLUGIN=ON -D PLUMED_MODE=shared -D LAMMPS_INSTALL_RPATH=ON -D BUILD_SHARED_LIBS=yes -D CMAKE_INSTALL_PREFIX=${deepmd_root} -D CMAKE_INSTALL_LIBDIR=lib -D CMAKE_INSTALL_FULL_LIBDIR=${deepmd_root}/lib ../cmake
make -j4
make install
where deepmd_route is $HOME/local
make -j8
make install
```

Now patching , (not sure of this step  )

`patchelf --add-rpath "/home/madhavan/anaconda3/envs/lpd_env/lob/python3.11/site-packages/tensorflow" `liblammps.so``

Any guidance will be be really helpful for my project. 

Thanking you ,

maddy

**Comment by njzjz** [comment]
> Additionally , no deepmd pairstyle present when I list the styles using , /home/madhavan/local/bin/lmp -h

`lmp -h` does not show plugins.

> I doubled checked by runnning a small lammps simulation (after loading the plugin
> `pluigin load /home/madhavan/Softwares/deepmd-kit/lb/libdeepmd_lmp.so` ) and an error saying unknown command plugin load.

I am unsure if you typed correctly, but it should be `plugin` instead of `pluigin`.

> I also tried setting the the LAMMPS_PLUGIN_PATH environment variable and run the small lammps simulation . This time, it showed no deepmd pair style.

Are there any errors on the screen?

**Comment by magnamad** [comment]
I am not sure if I did correctly then ,sorry. But I ran it the correct way now.  `plugin load /home/madhavan/Softwares/deepmd-kit/lb/libdeepmd_lmp.so ` There are errors on screen. Same  error when setting the  LAMMPS_PLUGIN_PATH environment variable and running.



```
LAMMPS (29 Sep 2021 - Update 2)
  using 2 OpenMP thread(s) per MPI task
Reading data file ...
  orthogonal box = (0.0000000 0.0000000 0.0000000) to (20.000000 20.000000 20.000000)
  2 by 2 by 2 MPI processor grid
  reading atoms ...
  62 atoms
  read_data CPU = 0.001 seconds
ERROR: Unrecognized pair style 'deepmd' (src/force.cpp:278)
Last command: pair_style      deepmd graph-compress.pb
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[62889,1],7]
  Exit code:    1
--------------------------------------------------------------------------

```

**Reply by njzjz** [reply]
There should be some messages outputted by `plugin` itself, but I don't see it.

**Reply by njzjz** [reply]
Like #1848

**Comment by magnamad** [comment]
Ah , sorry again. I missed the first line.

```
LAMMPS (29 Sep 2021 - Update 2)
  using 2 OpenMP thread(s) per MPI task
ERROR: Unknown command: plugin load /home/madhavan/Softwares/deepmd-kit/lib/libdeepmd_lmp.so (src/input.cpp:274)
Last command: plugin load /home/madhavan/Softwares/deepmd-kit/lib/libdeepmd_lmp.so
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[41782,1],0]
  Exit code:    1
--------------------------------------------------------------------------


```

**Reply by njzjz** [reply]
It looks to me that the `plugin` feature is not actually enabled.
https://docs.lammps.org/plugin.html

**Comment by magnamad** [comment]
Hi ,

I have enabled plugin during the build

```
cmake -D PKG_PLUGIN=ON -D PLUMED_MODE=shared -D LAMMPS_INSTALL_RPATH=ON -D BUILD_SHARED_LIBS=yes -D CMAKE_INSTALL_PREFIX=${deepmd_root} -D CMAKE_INSTALL_LIBDIR=lib -D CMAKE_INSTALL_FULL_LIBDIR=${deepmd_root}/lib ../cmake
make -j4
make install
where deepmd_route is $HOME/local
make -j8
make install
```

I am not sure of the patching step provided in the installation guide  , could that be the issue ?

`patchelf --add-rpath "/home/madhavan/anaconda3/envs/lpd_env/lob/python3.11/site-packages/tensorflow" liblammps.so`

**Reply by njzjz** [reply]
I don't see anything wrong here.

**Comment by njzjz** [comment]
You might use `export LD_DEBUG=libs` to debug which shared library is found and loaded.
